lib/datadog/appsec/contrib/devise/patcher.rb

```diff
@@ -0,0 +1,160 @@
+# frozen_string_literal: true
+
+require_relative '../patcher'
+require_relative 'resource'
+require_relative 'ext'
+require_relative '../../../kit/appsec/events'
+
+module Datadog
+  module AppSec
+    module Contrib
+      module Devise
+        AUTOMATED_USER_EVENT_TAGGING_BLOCK = proc do |trace, event|
+          trace.set_tag("_dd.appsec.events.#{event}.auto.mode", AppSec.settings.automated_track_user_events)
+        end
+
+        private_constant :AUTOMATED_USER_EVENT_TAGGING_BLOCK
+
+        # Hook in devise validate method
+        module AuthenticablePatch
+          # rubocop:disable Metrics/MethodLength, Metrics/PerceivedComplexity
+          def validate(resource, &block)
+            result = super
+            return result unless AppSec.enabled?
+            return result if AppSec.settings.automated_track_user_events == Ext::DISABLED_MODE
+
+            active_trace = defined?(Datadog::Tracing) && Datadog::Tracing.active_trace
+
+            return result unless active_trace
+
+            devise_resource = resource ? Resource.new(resource) : nil
+
+            event_information = {}
+            user_id = nil
+
+            if AppSec.settings.automated_track_user_events == Ext::EXTENDED_MODE && devise_resource
+              resource_email = devise_resource.email
+              resource_username = devise_resource.username
+
+              event_information[:email] = resource_email if resource_email
+              event_information[:username] = resource_username if resource_username
+            end
+
+            user_id = devise_resource.id if devise_resource && devise_resource.id
+
+            if result
+              if user_id
+                Datadog::Kit::AppSec::Events.track_login_success(
+                  active_trace,
+                  user: { id: user_id.to_s },
+                  **event_information,
+                  &AUTOMATED_USER_EVENT_TAGGING_BLOCK
+                )
```

Comment:
Hmm, this makes `Datadog::AppSec` depend on `Datadog::Kit`, creating a circular dependency.

The main idea behind `Datadog::Kit` is to provide users with a glue API for users to peruse. The dependency order should be:

`Datadog::Core` < `Datadog::{AppSec,Tracing,...}` < `Datadog::Kit`

IOW, components are implemented in terms of `Core`, and `Kit` is implemented atop components.

I think this is why this block argument looks so unwieldy.